src/memo/memo.ml

```diff
@@ -969,23 +965,47 @@ end = struct
       | Needs_work { sample_attempt; _ } -> Running sample_attempt
   end
 
-  let deps_changed =
-    let rec go deps =
-      match deps with
-      | [] -> Fiber.return Changed_or_not.Unchanged
-      | Last_dep.T (dep, v_id) :: deps -> (
-        match Exec_unknown.exec_dep_node_internal dep with
-        | Error cycle_error ->
-          Fiber.return (Changed_or_not.Cycle_error cycle_error)
-        | Ok res -> (
-          let* res = res in
-          match Value_id.equal res.id v_id with
-          | true -> go deps
-          | false -> Fiber.return Changed_or_not.Changed ) )
-    in
-    go
-
-  let do_validate (dep_node : _ Dep_node.t) ivar running_state =
+  let restore_from_cache (last_cached_value : _ Cached_value.t option) =
+    match last_cached_value with
+    | None -> Fiber.return (Error Cache_lookup_error.Not_found)
+    | Some cached_value -> (
+      match cached_value.value with
+      | Cancelled _dependency_cycle ->
+        (* Dependencies of cancelled computations are not accurate, so we can't
+           use [deps_changed] in this case. *)
+        Fiber.return (Error Cache_lookup_error.Not_found)
+      | Error _ ->
+        (* We always recompute errors, so there is no point in checking if any
+           of their dependencies changed. In principle, we could introduce
+           "persistent errors" that are recomputed only when their dependencies
+           have changed. *)
+        Fiber.return (Error Cache_lookup_error.Not_found)
+      | Ok _ -> (
+        let+ deps_changed =
+          let rec go deps =
+            match deps with
+            | [] -> Fiber.return Changed_or_not.Unchanged
+            | Last_dep.T (dep, v_id) :: deps -> (
+              match Exec_unknown.exec_dep_node_internal dep with
+              | Error dependency_cycle ->
+                Fiber.return (Changed_or_not.Cancelled { dependency_cycle })
+              | Ok res -> (
+                let* res = res in
```

Comment:
This confused me for a bit, maybe it's worth adding a comment. Something like:

(* A dependency cycle can totally be introduced and discovered during `res` computation, however it's reported as a normal error then. *)

(`res` sounds like a strange name for something that's barely even started, by the way)

Reply:
Fixed.